### PR TITLE
Add support for `X-Query-Tags`

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:

1. Now client can send any metadata about query via `X-Query-Tags` HTTP header
2. Also record `X-Query-Tags` in `metrics.go` for instant and range queries.

**Which issue(s) this PR fixes**:
Fixes: NA

**Special notes for your reviewer**:

Some context and rationale are discussed in https://github.com/grafana/loki/pull/4796#issuecomment-975547927
and [internal discussion](https://raintank-corp.slack.com/archives/C029V4SSS9L/p1637589440378600)

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/loki/modules.go [hunk 1]
```diff
@@ -494,6 +494,7 @@ func (t *Loki) initQueryFrontend() (_ services.Service, err error) {
 		queryrange.StatsHTTPMiddleware,
 		serverutil.NewPrepopulateMiddleware(),
 		serverutil.ResponseJSONMiddleware(),
+		serverutil.ExtractQueryTagsMiddleware(),
```
Comment (kavirajk) [inline pkg/loki/modules.go:497]: would be nice if I could lock this middleware chain for Query handler by some tests!. 

Also bit tricky to write one with all the dependency of `initQueryFrontend`.

Any thoughts are welcome!

File: pkg/logql/metrics.go [hunk 2]
```diff
@@ -98,6 +100,7 @@ func RecordMetrics(ctx context.Context, p Params, status string, stats stats.Res
 		"step", p.Step(),
 		"duration", time.Duration(int64(stats.Summary.ExecTime*float64(time.Second))),
 		"status", status,
+		"query_tags", queryTags,
```
Comment (cyriltovena) [inline pkg/logql/metrics.go:103]: I would actually try to output those as:

`status=success source=Value limit=1000`

instead of

`status=success query_tags=\"Source=logvolhist\" limit=1000`

File: pkg/util/server/middleware.go [hunk 3]
```diff
@@ -31,3 +32,17 @@ func ResponseJSONMiddleware() middleware.Interface {
 		})
 	})
 }
+
+func ExtractQueryTagsMiddleware() middleware.Interface {
+	return middleware.Func(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			ctx := req.Context()
+			tagsKey := "X-Query-Tags"
```
Comment (cyriltovena) [inline pkg/util/server/middleware.go:40]: A const would be better.

File: pkg/util/server/middleware.go [hunk 4]
```diff
@@ -31,3 +32,17 @@ func ResponseJSONMiddleware() middleware.Interface {
...
+			tags := req.Header.Get(tagsKey)
+			if tags != "" {
+				ctx = context.WithValue(ctx, tagsKey, tags)
+			}
+			req = req.WithContext(ctx)
```
Comment (cyriltovena) [inline pkg/util/server/middleware.go:45]: I think you can move this up.

File: pkg/logql/metrics.go (hunk 2 above)
Comment (kavirajk) [reply pkg/logql/metrics.go:103]: Ah. good point! @cyriltovena . I think I understand your concern!. But few questions.
1. You mean we don't need any differentiation that its `query_tags`? (e.g: If we just put `source=Value`, we may not know where this source comes from? From what header.
2. How do you want to log multiple values for `X-Query-Tags` header? (say `X-Query-Tags: Source=Foo,Feature=Beta`)
`status=success source=foo feature=beta`?

Comment (kavirajk) [reply pkg/logql/metrics.go]: @cyriltovena Fixed it!. It now logs in the desired format `status=success source=Value limit=1000`

File: pkg/util/server/middleware.go (hunk 4 above)
Comment (kavirajk) [reply pkg/util/server/middleware.go]: Done 👍

File: pkg/util/server/middleware.go (hunk 3 above)
Comment (kavirajk) [reply pkg/util/server/middleware.go]: Done 👍

File: pkg/logql/metrics.go [hunk 5]
```diff
@@ -88,20 +89,29 @@ func RecordMetrics(ctx context.Context, p Params, status string, stats stats.Res
 		returnedLines = int(result.(logqlmodel.Streams).Lines())
 	}
 
-	// we also log queries, useful for troubleshooting slow queries.
-	level.Info(logger).Log(
+	queryTags, _ := ctx.Value(serverutil.QueryTagsHTTPHeader).(string) // it's ok to be empty.
+
+	logValues := make([]interface{}, 0)
```
Comment (cyriltovena) [inline pkg/logql/metrics.go:94]:
```suggestion
	logValues := make([]interface{}, 0,20)
```

File: pkg/logql/metrics.go [hunk 6]
```diff
@@ -88,20 +89,29 @@ func RecordMetrics(ctx context.Context, p Params, status string, stats stats.Res
...
+	logValues = append(logValues, tagsToKeyValues(queryTags)...)
```
Comment (cyriltovena) [inline pkg/logql/metrics.go:95]: might be better to reverse the order and have tags at the end.

File: pkg/util/server/middleware.go [hunk 7]
```diff
@@ -1,12 +1,24 @@
...
+	"context"
 	"net/http"
+	"regexp"
 
 	"github.com/weaveworks/common/httpgrpc"
 	"github.com/weaveworks/common/middleware"
 )
 
+// NOTE(kavi): Why new type?
+// Our linter won't allow to use basic types like string to be used as key in context.
```
Comment (cyriltovena) [inline pkg/util/server/middleware.go:13]: It's normal this is the idiomatic way to use context key in go.

**Review by cyriltovena — APPROVED** [review]
LGTM

File: pkg/logql/metrics.go [hunk 8]
```diff
@@ -133,3 +144,34 @@ func QueryType(query string) (string, error) {
...
+// tagsToKeyValues converts QueryTags to form that is easy to log.
+// e.g: `Source=foo,Feature=beta` -> []interface{}{"source", "foo", "feature", "beta"}
+// so that we could log nicely!
+// If queryTags is not in canonical form then its completely ignored (e.g: `key1=value1,key2=value`)
+func tagsToKeyValues(queryTags string) []interface{} {
+	toks := strings.FieldsFunc(queryTags, func(r rune) bool {
+		return r == ','
+	})
+
+	vals := make([]string, 0)
```
Comment (jeschkies) [inline pkg/logql/metrics.go:157]: How about allocating `len(toks)`?

File: pkg/logql/metrics.go [hunk 9]
```diff
@@ -133,3 +144,34 @@ func QueryType(query string) (string, error) {
...
+
+	for _, tok := range toks {
+		val := strings.FieldsFunc(tok, func(r rune) bool {
+			return r == '='
+		})
+
+		if len(val) != 2 {
```
Comment (jeschkies) [inline pkg/logql/metrics.go:164]: Why is this here?

File: pkg/logql/metrics.go [hunk 10]
```diff
@@ -133,3 +144,34 @@ func QueryType(query string) (string, error) {
...
 }
+
+// tagsToKeyValues converts QueryTags to form that is easy to log.
+// e.g: `Source=foo,Feature=beta` -> []interface{}{"source", "foo", "feature", "beta"}
+// so that we could log nicely!
+// If queryTags is not in canonical form then its completely ignored (e.g: `key1=value1,key2=value`)
+func tagsToKeyValues(queryTags string) []interface{} {
+	toks := strings.FieldsFunc(queryTags, func(r rune) bool {
+		return r == ','
+	})
```
Comment (jeschkies) [inline pkg/logql/metrics.go:155]: Out of curiosity, why are you not using `Split`?

**Review by jeschkies — COMMENTED** [review]
Looks good to me except that I don't really understand where `QueryTagsHTTPHeader` will be used.

File: pkg/logql/metrics.go (hunk 9 above)
Comment (kavirajk) [reply pkg/logql/metrics.go:164]: To check our header value is in canonical form `key1=key2, key3=key4`. Once you split by =, the number of tokens should be 2.

File: pkg/logql/metrics.go (hunk 10 above)
Comment (kavirajk) [reply pkg/logql/metrics.go:155]: Split was a mistake (its behaviour is unintuitive), `Fields` is something usually you want to use (which was added later to stdlib. To avoid Go1 compatibility I think they put it in different name AFAIR)

File: pkg/logql/metrics.go (hunk 8 above)
Comment (kavirajk) [reply pkg/logql/metrics.go:157]: Good question! Actually we can't predict the len from toks.

Why?

Because tok may or may not be in canonical form e.g (`key=val') or (`abc`). 

if all the keys are in correct form, then you will have `2*len(toks)` but if any of them is invalid we have to skip them!. Hence difficult to predict!

File: pkg/logql/metrics.go (hunk 10 above)
Comment (kavirajk) [reply pkg/logql/metrics.go:155]: https://medium.com/@mlowicki/strings-fieldsfunc-vs-strings-split-96c667912f78

Comment (jeschkies) [reply pkg/logql/metrics.go:155]: Ah. Thanks!

File: pkg/logql/metrics.go (hunk 9 above)
Comment (jeschkies) [reply pkg/logql/metrics.go:164]: How about naming it `pair` or `keyValPair`?

File: pkg/logql/metrics.go (hunk 8 above)
Comment (jeschkies) [reply pkg/logql/metrics.go:157]: Fair point.

**Review by cyriltovena — APPROVED** [review]
LGTM

**Comment by pracucci** [comment]
This PR is creating some issues upgrading Loki in a downstream project (Cortex Enterprise Metrics) because, with the import changes in the PR, it leads to have GEM to also import `vendor/github.com/cortexproject/cortex/pkg/cortexpb/` which conflicts with other custom changes. Before this PR `vendor/github.com/cortexproject/cortex/pkg/cortexpb/` was not imported in GEM.

**Comment by pracucci** [comment]
I think the reason is that `pkg/util/server` (new import from `pkg/logql`) has `errors.go` which imports `github.com/grafana/loki/pkg/storage/chunk` which opens the pandora's box.
